libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt

```diff
@@ -214,41 +222,83 @@ internal class MapboxRerouteController @VisibleForTesting constructor(
         callback: NavigationRerouteController.RoutesCallback,
         routeOptions: RouteOptions
     ) {
-        requestId = directionsSession.requestRoutes(
-            routeOptions,
-            object : NavigationRouterCallback {
-                override fun onRoutesReady(
-                    routes: List<NavigationRoute>,
-                    routerOrigin: RouterOrigin
-                ) {
-                    mainJobController.scope.launch {
-                        state = RerouteState.RouteFetched(routerOrigin)
-                        state = RerouteState.Idle
-                        callback.onNewRoutes(routes, routerOrigin)
-                    }
-                }
+        rerouteJob = mainJobController.scope.launch {
+            val result = suspendCancellableCoroutine<RouteRequestResult> { cont ->
+                cont.invokeOnCancellation { onRequestInterrupted() }
+                requestId = directionsSession.requestRoutes(
+                    routeOptions,
+                    object : NavigationRouterCallback {
+                        override fun onRoutesReady(
+                            routes: List<NavigationRoute>,
+                            routerOrigin: RouterOrigin
+                        ) {
+                            if (cont.isActive) {
+                                cont.resume(RouteRequestResult.Success(routes, routerOrigin))
+                            }
+                        }
 
-                override fun onFailure(
-                    reasons: List<RouterFailure>,
-                    routeOptions: RouteOptions
-                ) {
+                        override fun onFailure(
+                            reasons: List<RouterFailure>,
+                            routeOptions: RouteOptions
+                        ) {
+                            if (cont.isActive) {
+                                cont.resume(RouteRequestResult.Failure(reasons))
+                            }
+                        }
+
+                        override fun onCanceled(
+                            routeOptions: RouteOptions,
+                            routerOrigin: RouterOrigin
+                        ) {
+                            if (cont.isActive) {
+                                cont.resume(RouteRequestResult.Cancellation)
+                            }
+                        }
+                    }
+                )
+            }
```

Comment:
maybe it should be extracted to a function `suspend fun requestRoute(routeOptions: RouteOptions): RouteRequestResult`?

Reply:
Sure.